cmd/main.go

```diff
@@ -372,14 +375,22 @@ func main() {
 
 	failOnError(autoMigrationWithLeader(autoMigrationLeader, db, log), "Failed auto migration process")
 
+	Options.UploaderConfig.AssistedServiceVersion = versions.GetRevision()
+	uploadClient := uploader.NewClient(&Options.UploaderConfig, db, log, ocpClient)
+
 	hostApi := host.NewManager(log.WithField("pkg", "host-state"), db, eventStreamWriter, eventsHandler, hwValidator,
 		instructionApi, &Options.HWValidatorConfig, metricsManager, &Options.HostConfig, lead, operatorsManager, providerRegistry, Options.EnableKubeAPI, objectHandler)
 	dnsApi := dns.NewDNSHandler(Options.BMConfig.BaseDNSDomains, log)
 	manifestsGenerator := network.NewManifestsGenerator(manifestsApi, Options.ManifestsGeneratorConfig)
 	clusterApi := cluster.NewManager(Options.ClusterConfig, log.WithField("pkg", "cluster-state"), db,
-		eventStreamWriter, eventsHandler, hostApi, metricsManager, manifestsGenerator, lead, operatorsManager, ocmClient, objectHandler, dnsApi, authHandler)
+		eventStreamWriter, eventsHandler, uploadClient, hostApi, metricsManager, manifestsGenerator, lead, operatorsManager, ocmClient, objectHandler, dnsApi, authHandler)
 	infraEnvApi := infraenv.NewManager(log.WithField("pkg", "host-state"), db, objectHandler)
 
+	clusterEventsUploader := thread.New(
+		log.WithField("pkg", "cluster-events-uploader"), "Cluster Events Uploader", Options.ClusterEventsUploaderInterval, clusterApi.UploadEvents)
```

Comment:
Why move this to a separate thread?

I don't see any huge problems with it, but I just generally don't like this pattern since we're not checking the health of this routine.

Reply:
On this note, what is the advantage of having a dedicated thread for this (or any other dedicated thread in this application) vs using a goroutine? We'll get thread overhead, but I can't see the benefit as the operations we launch in threads are not very resource-demanding - what am I missing?

Reply:
This came from a suggestion to prevent uploading data in the middle of a transaction.

Here's the background conversation: https://github.com/openshift/assisted-service/pull/4880#discussion_r1100512234

I understand your concerns, is there a better way to do this?

Reply:
Would including it in the health check along with the other monitor threads be a good idea to watch its health? https://github.com/openshift/assisted-service/blob/9b1ee70507b913eb4c3b90e288e85d304585ed85/cmd/main.go#L521

Reply:
Talked with @CrystalChun about this and it sounds like the reasoning was to ensure we're not sending events to an external system before committing the transaction that would actually transition the cluster to the new state.

We looked at some alternatives (like moving this to a post-transition function on the state machine), but those also run inside the same transaction (at least for cancel).

This is fine, but I think eventually we probably want to think about moving these threads out into their own container at some point in the future.

>what is the advantage of having a dedicated thread for this (or any other dedicated thread in this application) vs using a goroutine?

@rccrdpccl these are actually just goroutines in a poorly named package. This thread package is actually something we have internally to allow for looping logic. You can check it out [here](https://github.com/openshift/assisted-service/blob/f41f4b24d8773e90a816827c426925e00520ec96/pkg/thread/thread.go#L43)

Reply:
Thanks for the feedback @carbonin - it makes more sense now. My bad for not looking into its implementation :pray: